GH1517: add vswhere support

phillipsj opened the pull request:
<!--
BEFORE YOU CREATE A PULL REQUEST:

Ensure you have read over CONTRIBUTING - https://github.com/cake-build/cake/blob/develop/CONTRIBUTING.md. We provide VERY defined guidance (as such, we strongly adhere to it).

A summary of our expectations:
 - You are not submitting a pull request from your MASTER branch.
 - Do not reformat code, it makes it hard to see what has changed. Leave the formatting to us.

THANKS! We appreciate you reading the entire Contributing document and not just scanning through it.

DELETE EVERYTHING IN THIS COMMENT BLOCK
-->

phillipsj commented:
Fixes #1517

damianh commented on src/Cake.Common/Tools/VSWhere/VSWhereProduct.cs:
```diff
@@ -0,0 +1,32 @@
+    /// <summary>
+    /// Target Visual Studio product to path to find.
+    /// </summary>
+    public enum VSWhereProduct
```
`[Flags]` perhaps?

phillipsj commented:
Yeah I am planning to restructure it a lot more.  Flags is a good idea.

On Mar 6, 2017 3:37 PM, "Damian Hickey" <notifications@github.com> wrote:

> *@damianh* commented on this pull request.
> ------------------------------
>
> In src/Cake.Common/Tools/VSWhere/VSWhereProduct.cs
> <https://github.com/cake-build/cake/pull/1520#discussion_r104516170>:
>
> > @@ -0,0 +1,32 @@
> +﻿// Licensed to the .NET Foundation under one or more agreements.
> +// The .NET Foundation licenses this file to you under the MIT license.
> +// See the LICENSE file in the project root for more information.
> +
> +namespace Cake.Common.Tools.VSWhere
> +{
> +    /// <summary>
> +    /// Target Visual Studio product to path to find.
> +    /// </summary>
> +    public enum VSWhereProduct
>
> [Flags] perhaps?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cake-build/cake/pull/1520#pullrequestreview-25356956>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABfthVyJvXU9NGP_MpmFmZvWfLUqCkbXks5rjG6HgaJpZM4MTqVp>
> .
>

damianh commented on src/Cake.Common/Tools/VSWhere/VSWhereSettings.cs:
```diff
@@ -0,0 +1,56 @@
+        /// <summary>
+        /// Gets or sets s version range for instances to find. Example: ["15.0","16.0"] will find versions 15.*.
+        /// </summary>
+        public List<string> Version { get; set; }
```
Prob just a string? I think it's like nuget's version range (am not 100% sure). Does cake have a type for that if it is?

phillipsj commented:
@damianh I have some more cleanup to do, but I am following a similar pattern to Chocolatey and NuGet.

phillipsj commented on src/Cake.Common/Tools/NuGet/Add/NuGetAdder.cs:
```diff
@@ -21,7 +21,7 @@ public sealed class NuGetAdder : NuGetTool<NuGetAddSettings>
         /// </summary>
         /// <param name="fileSystem">The file system.</param>
-        /// <param name="environment">The environment.</param>
+        /// <param name="environment">The file system.</param>
```
Yep, nice catch.  I wonder how I managed that.

devlead replied:
@phillipsj No worries, just ping us, when you fixed, rebased and squashed commits and we'll do a review again.

phillipsj commented:
@devlead @damianh I have updated everything, rebased and squashed.

phillipsj commented:
@devlead PR has been updated and now includes the legacy command.

devlead commented on src/Cake.Common/Tools/VSWhere/VSWhereAliases.cs:
```diff
@@ -0,0 +1,246 @@
+        [CakeMethodAlias]
+        [CakeAliasCategory("Legacy")]
+        [CakeNamespaceImport("Cake.Common.Tools.VSWhere.Legacy")]
+        public static string VSLegacy(this ICakeContext context, bool latest)
```
Should return FilePath

devlead replied:
What's the definition/purpose latest? And is 2017 already legacy?

Running 
```cake
Information("VSLegacy(true): {0}", VSLegacy(true));
Information("VSLegacy(false): {0}", VSLegacy(false));
Information("VSLegacy(VSWhereLegacySettings): {0}", string.Join("\r\n\t", VSLegacy(new VSWhereLegacySettings{})));
```
outputs  the following
```
VSLegacy(true): C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise
VSLegacy(false): C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise
VSLegacy(VSWhereLegacySettings): C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise
        D:\VisualStudio\Microsoft Visual Studio 14.0\
```

phillipsj replied:
The purpose according to the documentation is to return the newest version or last installed and includes 2015 or older instead of just being limited to VS2017.  Latest only searches 2017 from my understanding.

devlead replied:
Ah ok 👍

devlead commented on src/Cake.Common/Tools/VSWhere/VSWhereAliases.cs:
```diff
@@ -0,0 +1,246 @@
+        [CakeMethodAlias]
+        [CakeAliasCategory("Latest")]
+        [CakeNamespaceImport("Cake.Common.Tools.VSWhere.Latest")]
+        public static string VSLatest(this ICakeContext context)
```
Should return FilePath

devlead commented on src/Cake.Common/Tools/VSWhere/VSWhereAliases.cs:
```diff
@@ -0,0 +1,246 @@
+        [CakeMethodAlias]
+        [CakeAliasCategory("Latest")]
+        [CakeNamespaceImport("Cake.Common.Tools.VSWhere.Latest")]
+        public static string VSLatest(this ICakeContext context, VSWhereLatestSettings settings)
```
Should return FilePath

devlead commented on src/Cake.Common/Tools/VSWhere/VSWhereAliases.cs:
```diff
@@ -0,0 +1,246 @@
+        [CakeMethodAlias]
+        [CakeAliasCategory("Legacy")]
+        [CakeNamespaceImport("Cake.Common.Tools.VSWhere.Legacy")]
+        public static IEnumerable<string> VSLegacy(this ICakeContext context, VSWhereLegacySettings settings)
```
Should return `IEnumerable<FilePath>`

devlead commented on src/Cake.Common/Tools/VSWhere/VSWhereAliases.cs:
```diff
@@ -0,0 +1,246 @@
+        [CakeMethodAlias]
+        [CakeAliasCategory("All")]
+        [CakeNamespaceImport("Cake.Common.Tools.VSWhere.All")]
+        public static IEnumerable<string> VSAll(this ICakeContext context)
```
Should return `IEnumerable<FilePath>`

devlead commented on src/Cake.Common/Tools/VSWhere/VSWhereAliases.cs:
```diff
@@ -0,0 +1,246 @@
+        [CakeMethodAlias]
+        [CakeAliasCategory("All")]
+        [CakeNamespaceImport("Cake.Common.Tools.VSWhere.All")]
+        public static IEnumerable<string> VSAll(this ICakeContext context, VSWhereAllSettings settings)
```
Should return `IEnumerable<FilePath>`

gep13 replied:
Why did we go from VSWhereAll to VSAll?  Same for the other aliases.  I would have thought that the former would make sense.

devlead commented on src/Cake.Common/Tools/VSWhere/VSWhereAliases.cs:
```diff
@@ -0,0 +1,246 @@
+        [CakeMethodAlias]
+        [CakeAliasCategory("Product")]
+        [CakeNamespaceImport("Cake.Common.Tools.VSWhere.Product")]
+        public static IEnumerable<string> VSProducts(this ICakeContext context)
```
Should return IEnumerable<FilePath>

devlead replied:
This fails for we with exit code 87 (`Error 0x57: Argument required for: products`)
Docs state
```
One or more products to find. Defaults to Community, Professional, and Enterprise.
Specify "*" by itself to search all product instances installed.
``` 
Verified by calling ` -products * -property installationPath -nologo` it doesn't get this error, so if no value specified it should fall back to `*`

devlead commented on src/Cake.Common/Tools/VSWhere/VSWhereAliases.cs:
```diff
@@ -0,0 +1,246 @@
+        [CakeMethodAlias]
+        [CakeAliasCategory("Product")]
+        [CakeNamespaceImport("Cake.Common.Tools.VSWhere.Product")]
+        public static IEnumerable<string> VSProducts(this ICakeContext context, string products)
```
Should return `IEnumerable<FilePath>`

devlead replied:
Wondering if `string products` shouldn't be `IEnumerable<string>` looking at their examples it seems to be a space separated string
```
set ids=%pre%Community %pre%Professional %pre%Enterprise %pre%BuildTools
vswhere -latest -products %ids%
```

devlead commented on src/Cake.Common/Tools/VSWhere/VSWhereAliases.cs:
```diff
@@ -0,0 +1,246 @@
+        [CakeMethodAlias]
+        [CakeAliasCategory("Product")]
+        [CakeNamespaceImport("Cake.Common.Tools.VSWhere.Product")]
+        public static IEnumerable<string> VSProducts(this ICakeContext context, string products, VSWhereProductSettings settings)
```
Should return `IEnumerable<FilePath>`

devlead commented on src/Cake.Common/Tools/VSWhere/Product/VSWhereProductSettings.cs:
```diff
@@ -0,0 +1,17 @@
+        /// <summary>
+        /// Gets or sets the products to find. Defaults to Community, Professional, and Enterprise. Specify "*" by itself to search all product instances installed.
+        /// </summary>
+        public string Products { get; set; }
```
This will currently be ignored if set see https://github.com/cake-build/cake/pull/1520/files#diff-96c402755256a225c5f4eb14a5f2586bR241

devlead commented on src/Cake.Common/Tools/VSWhere/VSWhereAliases.cs:
```diff
@@ -0,0 +1,246 @@
+            {
+                throw new ArgumentNullException(nameof(settings));
+            }
+            settings.Products = products;
```
I'm thinking either drop `string products` or make `VSWhereProductSettings.Products` internal.

gep13 commented on src/Cake.Common/Tools/VSWhere/VSWhereAliases.cs:
```diff
@@ -0,0 +1,246 @@
+namespace Cake.Common.Tools.VSWhere
+{
+    /// <summary>
+    /// Contains functionality related to executing VSWhere tool.
```
I think we should have information here about requiring the installation of a tool in order for this to work, unless I have missed something.  Similar to what we have done here:

![image](https://cloud.githubusercontent.com/assets/1271146/23920639/6632989e-08f3-11e7-9c13-5f853a12554a.png)

phillipsj commented:
@devlead @gep13. Awesome feedback.  If I get some energy tonight I will try to get those changes made.

heaths commented on src/Cake.Common/Tools/VSWhere/VSWhereTool.cs:
```diff
@@ -0,0 +1,86 @@
+                builder.AppendQuoted(settings.Requires);
+            }
+
+            if (!string.IsNullOrWhiteSpace(settings.ReturnProperty))
```
While you default this to "installationPath", if someone overrides it to be null or empty, the output would otherwise be unprocessed and, by default, in "name: value" form. You don't want to map that to any sort of key/value pair and just return it raw? (Perfectly fine, just calling it out in case it was unintended.)

phillipsj commented:
I thought about that.  I just made a decision to get the conversation
started

On Mar 14, 2017 4:59 PM, "Heath Stewart" <notifications@github.com> wrote:

*@heaths* commented on this pull request.
------------------------------

In src/Cake.Common/Tools/VSWhere/VSWhereTool.cs
<https://github.com/cake-build/cake/pull/1520#discussion_r106025718>:

> +        /// <returns>The process argument builder.</returns>
+        protected ProcessArgumentBuilder
AddCommonArguments(VSWhereSettings settings, ProcessArgumentBuilder
builder)
+        {
+            if (!string.IsNullOrWhiteSpace(settings.Version))
+            {
+                builder.Append("-version");
+                builder.AppendQuoted(settings.Version);
+            }
+
+            if (!string.IsNullOrWhiteSpace(settings.Requires))
+            {
+                builder.Append("-requires");
+                builder.AppendQuoted(settings.Requires);
+            }
+
+            if (!string.IsNullOrWhiteSpace(settings.ReturnProperty))

While you default this to "installationPath", if someone overrides it to be
null or empty, the output would otherwise be unprocessed and, by default,
in "name: value" form. You don't want to map that to any sort of key/value
pair and just return it raw? (Perfectly fine, just calling it out in case
it was unintended.)

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/cake-build/cake/pull/1520#pullrequestreview-26930636>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/ABfthbttbuiMjJP2yCfY2QkmobmtLagyks5rlv_NgaJpZM4MTqVp>
.

patriksvensson commented on src/Cake.Common.Tests/Unit/Tools/VSWhere/All/VSWhereAllTests.cs:
```diff
@@ -0,0 +1,174 @@
+
+            [WindowsTheory]
+            [InlineData("C:/vswhere/vswhere.exe", "C:/vswhere/vswhere.exe")]
+            public void Should_Use_VSWhere_Executable_From_Tool_Path_If_Provided_On_Windows(string toolPath, string expected)
```
Since it's only a single case, this could probably be converted to a `WindowsFact`.

patriksvensson commented on src/Cake.Common.Tests/Unit/Tools/VSWhere/Latest/VSWhereLatestTests.cs:
```diff
@@ -0,0 +1,174 @@
+
+            [WindowsTheory]
+            [InlineData("C:/vswhere/vswhere.exe", "C:/vswhere/vswhere.exe")]
+            public void Should_Use_VSWhere_Executable_From_Tool_Path_If_Provided_On_Windows(string toolPath, string expected)
```
Since it's only a single case, this could probably be converted to a `WindowsFact`.

patriksvensson commented on src/Cake.Common.Tests/Unit/Tools/VSWhere/Legacy/VSWhereLegacyTests.cs:
```diff
@@ -0,0 +1,175 @@
+
+            [WindowsTheory]
+            [InlineData("C:/vswhere/vswhere.exe", "C:/vswhere/vswhere.exe")]
+            public void Should_Use_VSWhere_Executable_From_Tool_Path_If_Provided_On_Windows(string toolPath, string expected)
```
Since it's only a single case, this could probably be converted to a `WindowsFact`.

patriksvensson commented on src/Cake.Common.Tests/Unit/Tools/VSWhere/Product/VSWhereProductTests.cs:
```diff
@@ -0,0 +1,188 @@
+            }
+
+            [WindowsTheory]
+            [InlineData("C:/vswhere/vswhere.exe", "C:/vswhere/vswhere.exe")]
```
Since it's only a single case, this could probably be converted to a `WindowsFact`.

phillipsj replied:
@patriksvensson That throws and error saying inline data is not allowed for a Fact.  I am more than happy to make the change and just put it directly in the test.  It is like that in other parts of the code, an example would be the nuget tests, so that is why I did it that way.

heaths commented on src/Cake.Common/Tools/VSWhere/Product/VSWhereProduct.cs:
```diff
@@ -55,6 +56,10 @@ private ProcessArgumentBuilder GetArguments(VSWhereProductSettings settings)
             }
+            else
+            {
+                builder.AppendQuoted("*");
```
This is not recommended. The default is Community, Professional, and Enterprise. Doing this you say "any product". The point was this is **vs**where and primarily geared for VS. If they want more, they should pass it in. Defaulting to "any product" really alters the behavior for anyone used to how vswhere.exe works.

devlead replied:
Calling without * causes a fail for we with exit code 87 (Error 0x57: Argument required for: products)

phillipsj replied:
I also get the same error with the newest version.

![image](https://cloud.githubusercontent.com/assets/1568133/24018981/8dd4eef4-0a6c-11e7-977d-b749510c2d47.png)

heaths replied:
Well, yes, if you specify `-products`. But if they didn't pass a value, don't specify `-products`. An argument is required only when the parameter is specified; otherwise, it defaults to:

* Microsoft.VisualStudio.Product.Community
* Microsoft.VisualStudio.Product.Professional
* Microsoft.VisualStudio.Product.Enterprise

phillipsj commented:
@devlead @gep13 @heaths I have made more changes, and squashed it all.  I removed the ability to run the products flag without passing in a value for products.

phillipsj commented:
Just going to ping everyone again.

heaths commented on src/Cake.Common.Tests/Unit/Tools/VSWhere/Product/VSWhereProductTests.cs:
```diff
@@ -0,0 +1,188 @@
+                var result = fixture.Run();
+
+                // Then
+                Assert.Equal("-products \"BuildTools\" -property installationPath -nologo", result.Args);
```
While this is fine for tests, keep in mind the full product name is "Microsoft.VisualStudio.Product.BuildTools". I have a PR to support short names but haven't merged it yet because of no demand. It would also increase the size of the EXE a bit more.

heaths commented on src/Cake.Common/Tools/VSWhere/Product/VSWhereProduct.cs:
```diff
@@ -0,0 +1,73 @@
+        {
+            var builder = new ProcessArgumentBuilder();
+
+            builder.Append("-products");
```
This doesn't look fixed. "-products" should only be appended if `settings.Products` is non-null. Move this into the following condition block and eliminate the `else`.

devlead replied:
@phillipsj  could you look at sorting this and I'll do a final review before merging this.

phillipsj replied:
Will do.

phillipsj replied:
@heaths @devlead Changes have been made.

devlead commented:
@phillipsj  your changes have been merged, thanks for your contribution 👍

gep13 commented:
@phillipsj prior to the next release, would you be in a position to write a short blog post about this new Alias?  Showing where it would be used, etc?
